### DISCUSSION TITLE
How to upload to SD card in ESP3D 3.0?

### DISCUSSION THREAD

**Discussion by Stephano2000** [luc-github/ESP3D#925] [F.A.Q] [answered]
It has been some time since i haven't used ESP3D and Marlin, so i might have oddly forgotten some basics.
ESP3D 3.0
Ramps board + 12864 LCD controller with SD card
ESP3D web interface shows the contents of the SD card along with run and delete buttons. However there is no upload button. The dropdown menu shows only refresh.
I remember i used to upload to the SD card, but this could have been on ESP3D 2.x.
I have uncommented BINARY_FILE_TRANSFER.

Any pointers?
Thank you

**Comment by luc-github** [comment, marked as answer]
Supported SD modes in V3 are described here : http://esp3d.io/esp3d/v3.x/documentation/sdtransfer/index.html
and status is here : https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-2897564

```
UPLOAD_SERIAL_SD (implemented in 2.1 but not more available in 3.0)
M28/M29 protocol is too slow and buggy with external queries
``` 

The M28/M29 protocol support has been dropped in V3

**Comment by Stephano2000** [comment]
I am currently researching this issue.

1- I understand that Marlin has an [optimized binary transfer mode](https://marlinfw.org/docs/gcode/M028.html) triggered by M28 B1, provided BINARY_FILE_TRANSFER is uncommented in Configuration_adv.h. Is this mode buggy and/or as well?
https://marlinfw.org/docs/gcode/M028.html

2- How is UPLOAD_FAST_SERIAL_SD set up with a ramps board with Marlin firmware?

Note: I am currently testing ESP3D 3.0 on an ESP-01 (so no extra GPIO pins)

**Comment by luc-github** [comment]
1 - M28 mode itself is not buggy, but subject to any external queries coming, which make it not reliable, and it is slow like hell ,so after using it once, people do not use anymore.
M28 B1 is an experimental protocol that is subject to change, so I won't work on it, knowing it the only benefit is speed (may be), but not reliability, so same problem as normal M28 protocol

2 - you cannot, it is used only by MKS Marlin code and linked to specific TFT in Marlin - not related to ESP3D

**Comment by Stephano2000** [comment]
1- Is it possible to restore the slow and problematic M28/M29 mode for ESP3D v3? Transferring small gcode test files back and forth via SD card when doing printer calibration and testing isn't fun.

2- Maybe the wording isn't clear then:
 [UPLOAD_FAST_SERIAL_SD](https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-2897564): "it is currently limited to MKS TFT **and boards using Marlin**"
Although "implemented for MKS" is stated, but it's still a little ambiguous. At least it was for me.

**Reply by luc-github** [reply]
1 - if you want to push a PR - I can consider  to merge it - but I won't spend time on this 
2 - what is ambiguous ? MKS board only and Marlin only

**Reply by andyspoo** [reply]
I was disappointing with the lack of upload as well, but I can completely see why luc-github has dropped it.
But after doing a quick search there are a couple of other possibilities that would make it redundant, i.e. you can buy a wifi memory cards from aliexpress , e.g. :
- ezshare Wireless WIFI SD Card which is about 25 dollars or so,
- BIGTREETECH BTT TF Cloud V1.0 SD Cloud V1.0 Wireless Transmission Module which is less than 10 dollars!!
I did not even know the BTT Cloud cards existed. I'm definately going to give that a try!

**Reply by luc-github** [reply]
@andyspoo be sure the ezshare work in upload and not in download  only- because it was the case years ago 
About altermative solution there is a topic : https://github.com/luc-github/ESP3D/discussions/535

about TF Cloud the BTT solution / Fysect solution : I was never able to make them work reliablely https://github.com/luc-github/ESP3D/issues/454#issuecomment-667010720, it is a hack and I do not support it neither because of reliability issue and user cases not handled , and I am not the only one  who got such result, the lastest version of Fysetc WiFi pro does not use this hack and now it is reliable: https://github.com/FYSETC/SD-WIFI-PRO,

**Reply by andyspoo** [reply]
Thanks for the info. This is an interesting video from Thomas Sanladerer = https://www.youtube.com/watch?v=nHNZPRl8gzA
